src/systems/filecoin_mining/storage_mining/storage_miner_actor_code.go

```diff
@@ -333,61 +244,247 @@ func (a *StorageMinerActorCode_I) _onSuccessfulPoSt(rt Runtime, postSubmission p
 		newTerminatedFaults,
 	)
 
-	a._submitPowerReport(rt, lastPoSt)
+	a._submitPowerReport(rt, lastPoStResponse)
 
 	// TODO: check EnsurePledgeCollateralSatisfied
 	// pledgeCollateralSatisfied
 
+	// Now that all is done update pointer to last response
 	h, st = a.State(rt)
-	st.ChallengeStatus().Impl().OnChallengeResponse(rt.CurrEpoch())
+	st.ChallengeStatus().Impl().OnPoStSuccess(rt.CurrEpoch())
 	st._processStagedCommittedSectors(rt)
 	UpdateRelease(rt, h, st)
 
 	return rt.SuccessReturn()
 
 }
 
-func (a *StorageMinerActorCode_I) SubmitSurprisePoSt(rt Runtime, postSubmission poster.PoStSubmission) InvocOutput {
+// called by verifier to update miner state on successful surprise post
+func (a *StorageMinerActorCode_I) SubmitSurprisePoSt(rt Runtime, onChainInfo sector.OnChainPoStVerifyInfo) InvocOutput {
 	TODO() // TODO: validate caller
 
+	// Verify correct PoSt Submission
+	isPoStVerified := a._verifySurprisePoSt(rt, onChainInfo)
+	if !isPoStVerified {
+		// no state transition, just error out and miner should submitSurprisePoSt again
+		// TODO: determine proper error here and error-handling machinery
+		rt.Abort("TODO")
+	}
+
+	return a._onSuccessfulPoSt(rt, onChainInfo)
+
+}
+
+func (a *StorageMinerActorCode_I) _verifySurprisePoSt(rt Runtime, onChainInfo sector.OnChainPoStVerifyInfo) bool {
+
+	// 1. Check that the miner in question is currently being challenged
 	if !a._isChallenged(rt) {
 		// TODO: determine proper error here and error-handling machinery
 		rt.Abort("cannot SubmitSurprisePoSt when not challenged")
 	}
 
-	// Verify correct PoSt Submission
-	isPoStVerified := a._verifyPoStSubmission(rt, postSubmission)
-	if !isPoStVerified {
-		// no state transition, just error out and miner should submitSurprisePoSt again
+	// to pull in from constants
+	PROVING_PERIOD := block.ChainEpoch(0)
+
+	// 2. Check that the challenge has not expired
+	// Check that miner can still submit (i.e. that the challenge window has not passed)
+	// This will prevent miner from submitting a Surprise PoSt past the challenge period
+	_, st := a.State(rt)
+	if rt.CurrEpoch() < st.ChallengeStatus().LastChallengeEpoch()+PROVING_PERIOD {
+		rt.Abort("cannot SubmitSurprisePoSt late")
```

Comment:
How about just returning `false`? This case doesn't seem like it should be particularly privileged.